rfd/xxxx-ssl-key-log.md

```diff
@@ -0,0 +1,30 @@
+---
+authors: Andrew Burke (andrew.burke@goteleport.com)
+state: draft
+---
+
+# RFD X - SSL Key Logging
+
+## What
+Teleport will optionally log TLS session keys. See [curl](https://everything.curl.dev/usingcurl/tls/sslkeylogfile) or [Firefox](https://firefox-source-docs.mozilla.org/security/nss/legacy/key_log_format/index.html) for examples of this implemented elsewhere.
+
+### Related issues
+- [#12295](https://github.com/gravitational/teleport/issues/12995)
+
+## Why
+Logging session keys allows users to inspect Teleport traffic with tools such as Wireshark.
+
+## Details
+Key logging will apply in the following scenarios:
+- All `tsh` clients
+- All `tctl` clients
+- Web API listener on proxies
+- Auth API listener on auth servers
+
+To enable keylogging, Teleport will use a new build flag, `sslkeylog`, enabled by setting the environment variable `ALLOW_SSLKEYLOGFILE` when compiling. When this build of Teleport is run and the `SSLKEYLOGFILE` environment variable is present, Teleport will open a file at `SSLKEYLOGFILE` and add it to the `tls.Config` of any client/listener than needs it (note: this means that the clients/listeners will be sharing a file handle. The key log file will be in append mode and calls to `os.File.Write` are atomic, so there should be no concurrency issues).
```

Comment:
What happens when this file grows? Is there any log rotation policy in place? Could this log to syslog as a separate service, something like `teleport_tls_keylog`?

I get this is inherently insecure and probably shouldn't be run in production settings, but I can see someone misusing this and I don't think mitigating it is too difficult.

Looking at `tls.Config`:

```
// KeyLogWriter optionally specifies a destination for TLS master secrets
	// in NSS key log format that can be used to allow external programs
	// such as Wireshark to decrypt TLS connections.
	// See https://developer.mozilla.org/en-US/docs/Mozilla/Projects/NSS/Key_Log_Format.
	// Use of KeyLogWriter compromises security and should only be
	// used for debugging.
	KeyLogWriter [io](https://pkg.go.dev/io).[Writer](https://pkg.go.dev/io#Writer)
```

Could we just set that to a syslog logger instead?

All of that said, since this is a compile opt in feature, I can accept "out of scope, if someone is using this feature, they should be aware of the risks"

Reply:
The problem is that the consumers of the key log file (wireshark) are looking for just a single log file, and log rotation will break that.